File: jar-common/src/test/java/com/beancounter/common/TestCurrency.java
package com.beancounter.common;

import static org.assertj.core.api.Assertions.assertThat;

import com.beancounter.common.contracts.CurrencyResponse;
import com.beancounter.common.model.Currency;
import com.beancounter.common.model.IsoCurrencyPair;
import com.beancounter.common.utils.CurrencyUtils;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collection;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;


class TestCurrency {
  private final CurrencyUtils currencyUtils = new CurrencyUtils();

  @Test
  void is_CurrencySerializing() throws Exception {

    Currency currency = Currency.builder()
        .code("SomeCode")
        .name("Some Name")
        .symbol("$")
        .build();

    assertThat(currency).isNotNull();

    ObjectMapper om = new ObjectMapper();
    String json = om.writeValueAsString(currency);

    Currency fromJson = om.readValue(json, Currency.class);

    assertThat(fromJson).isEqualTo(currency);
  }

  @Test
  void is_CurrencyResponseSerializing() throws Exception {

    CurrencyResponse currencyResponse = CurrencyResponse.builder().build();
    Collection<Currency> currencies = new ArrayList<>();
    Currency currency = Currency.builder()
        .code("SomeId")
        .name("Some Name")
        .symbol("$")
        .build();

    currencies.add(currency);
    currencyResponse.setData(currencies);

    ObjectMapper om = new ObjectMapper();
    String json = om.writeValueAsString(currencyResponse);

    CurrencyResponse fromJson = om.readValue(json, CurrencyResponse.class);

    assertThat(fromJson).isEqualTo(currencyResponse);
  }

  @Test
  void is_NullCurrencyCodeWorking() {
    Assertions.assertThrows(NullPointerException.class, ()
        -> Currency.builder().code(null).build());
  }

  @Test
  void is_GetCurrencyWorking() {
    Currency currency = currencyUtils.getCurrency("NZD");
    assertThat(currency).hasFieldOrPropertyWithValue("code", "NZD");
    assertThat(currencyUtils.getCurrency(null)).isNull();
  }

  @Test
  void is_CurrencyPairConsistent() {
    String trade = "NZD";
    String report = "USD";
    IsoCurrencyPair byCode = IsoCurrencyPair.builder()
        .from(report)
        .to(trade)
        .build();

    IsoCurrencyPair byCurrency = IsoCurrencyPair.from(
        currencyUtils.getCurrency(report), currencyUtils.getCurrency(trade));
    assertThat(byCode).isEqualToComparingFieldByField(byCurrency);

    assertThat(IsoCurrencyPair.from(
        currencyUtils.getCurrency(report), currencyUtils.getCurrency(report)))
        .isNull();

  }

  @Test
  void is_NullCurrencyAssertions() {

    assertThat(IsoCurrencyPair.from(
        currencyUtils.getCurrency("USD"), null))
        .isNull();
    assertThat(IsoCurrencyPair.from(
        null, currencyUtils.getCurrency("USD")))
        .isNull();
    assertThat(currencyUtils
        .getCurrencyPair(BigDecimal.ONE,
            currencyUtils.getCurrency("USD"), null))
        .isNull();
    assertThat(currencyUtils.getCurrencyPair(BigDecimal.ONE,
        null, currencyUtils.getCurrency("USD")))
        .isNull();

  }


  @Test
  void is_CorrectCurrencyPair() {
    assertThat(currencyUtils.getCurrencyPair(null, null, null)).isNull();

    assertThat(currencyUtils.getCurrencyPair(BigDecimal.TEN,
        currencyUtils.getCurrency("NZD"),
        currencyUtils.getCurrency("USD")))
        .isNull(); // We have a rate, so don't request one

    assertThat(currencyUtils.getCurrencyPair(null,
        currencyUtils.getCurrency("USD"), currencyUtils.getCurrency("USD")))
        .isNull();

    assertThat(currencyUtils.getCurrencyPair(null,
        currencyUtils.getCurrency("NZD"), currencyUtils.getCurrency("USD")))
        .isNotNull()
        .hasFieldOrPropertyWithValue("from", "NZD")
        .hasFieldOrPropertyWithValue("to", "USD");

  }

}

File: jar-common/src/test/kotlin/com/beancounter/common/TestCurrency.kt
package com.beancounter.common

import com.beancounter.common.contracts.CurrencyResponse
import com.beancounter.common.model.Currency
import com.beancounter.common.model.IsoCurrencyPair
import com.beancounter.common.model.IsoCurrencyPair.Companion.toPair
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

/**
 * Test suite for Currency model to ensure proper serialization, deserialization, and behavior.
 *
 * This class tests:
 * - Currency object serialization and deserialization
 * - Currency response handling
 * - Currency pair consistency
 * - Default value handling
 * - Currency contract compliance
 */
internal class TestCurrency {
    @Test
    fun `should serialize and deserialize currency correctly`() {
        val currency = TestHelpers.createTestCurrency("SomeCode")
        TestHelpers.assertSerializationRoundTrip(currency)
    }

    @Test
    fun `should serialize and deserialize currency response correctly`() {
        val currencies: MutableCollection<Currency> = ArrayList()
        val currency = TestHelpers.createTestCurrency("SomeId")
        currencies.add(currency)
        val currencyResponse = CurrencyResponse(currencies)
        TestHelpers.assertSerializationRoundTrip(currencyResponse)
    }

    @Test
    fun `should serialize and deserialize currency collection correctly`() {
        val currencies =
            listOf(
                TestHelpers.createTestCurrency("USD"),
                TestHelpers.createTestCurrency("EUR"),
                TestHelpers.createTestCurrency("GBP")
            )
        TestHelpers.assertCollectionSerializationRoundTrip(
            collection = currencies,
            typeReference = object : com.fasterxml.jackson.core.type.TypeReference<Collection<Currency>>() {}
        )
    }

    @Test
    fun `should create currency with correct code`() {
        val currency = TestHelpers.createTestCurrency("NZD")
        assertThat(currency).hasFieldOrPropertyWithValue(
            "code",
            "NZD"
        )
    }

    @Test
    fun `should maintain currency pair consistency`() {
        val trade = "NZD"
        val report = "USD"
        val byCode =
            IsoCurrencyPair(
                report,
                trade
            )
        val byCurrency =
            toPair(
                TestHelpers.createTestCurrency(report),
                TestHelpers.createTestCurrency(trade)
            )
        assertThat(byCode).usingRecursiveComparison().isEqualTo(byCurrency)
        assertThat(
            toPair(
                TestHelpers.createTestCurrency(report),
                TestHelpers.createTestCurrency(report)
            )
        ).isNull()
    }

    @Test
    fun `should handle default values correctly`() {
        val nzd = TestHelpers.createTestCurrency("NZD")
        assertThat(nzd).hasNoNullFieldsOrProperties()

        // Mutable
        nzd.name = "Aotearoa"
        nzd.symbol = "%"

        assertThat(nzd)
            .hasFieldOrPropertyWithValue(
                "name",
                "Aotearoa"
            ).hasFieldOrPropertyWithValue(
                "symbol",
                "%"
            )
    }

    @Test
    fun `should honor currency contract`() {
        val nzd = Currency("NZD")
        val sgd = Currency("SGD")
        assertThat(nzd).isNotEqualTo(sgd)
        val currencies = HashMap<Currency, Currency>()
        currencies[nzd] = nzd
        currencies[sgd] = sgd
        assertThat(currencies)
            .hasSize(2)
            .containsKeys(
                nzd,
                sgd
            )
    }

    @Test
    fun `should honor ISO pair contract`() {
        val pairA =
            IsoCurrencyPair(
                "USD",
                "NZD"
            )
        val pairB =
            IsoCurrencyPair(
                "USD",
                "NZD"
            )
        val pairs = HashMap<IsoCurrencyPair, IsoCurrencyPair>()
        pairs[pairA] = pairA
        pairs[pairB] = pairB
        assertThat(pairs).hasSize(1).containsKeys(
            pairA,
            pairB
        )
    }
}